### PR TITLE
Add mypy optional error codes

### PR DESCRIPTION
**Pull request by stinodego**
Mypy has some more [optional checks](https://mypy.readthedocs.io/en/stable/error_code_list2.html#check-that-expression-is-redundant-redundant-expr) that can help improve code quality.

I enabled two of them, and fixed some resulting errors.

Changes:
* Enabled [`truthy-bool` error code](https://mypy.readthedocs.io/en/stable/error_code_list2.html#check-that-expression-is-not-implicitly-true-in-boolean-context-truthy-bool) for mypy. We were already compliant.
* Enabled [`redundant-expr` error code](https://mypy.readthedocs.io/en/stable/error_code_list2.html#check-that-expression-is-redundant-redundant-expr) for mypy
  * `_ensure_datetime` had a redundant check; this has been fixed
  * `LazyFrame.with_columns` supports parsing a string as a literal. This was absent from the typing and docs. I added this.
  * `read_csv` had a redundant check for making sure columns are strings - this is ensured by the `handle_projection_columns`. This has been fixed
  * `assert_series_equal` checks if the inputs are indeed Series. This is redundant as the types are specified, but makes sense to include regardless. This error has been ignored.

### CODE REVIEW

**Comment by ritchie46** [comment]
> Enabled [redundant-expr error code](https://mypy.readthedocs.io/en/stable/error_code_list2.html#check-that-expression-is-redundant-redundant-expr) for mypy

That's a nice lint.

**Comment by ritchie46** [comment]
I request a review from @zundertj and will do vice versa his PR's as you two are far more knowledgeable about mypy than I am. And are stricter pythonistas ;)

File: py-polars/polars/io.py [hunk 1]
```diff
@@ -209,7 +209,7 @@ def read_csv(
 
     if columns and not has_header:
         for column in columns:
-            if isinstance(column, str) and not column.startswith("column_"):
+            if not column.startswith("column_"):
```
Comment (zundertj) [inline py-polars/polars/io.py:212]: `columns` is, before this statement, latest touched by `handle_projection_columns`, and can be a list of ints. Thus we cannot assume that `column` is always a string?

Comment (stinodego) [reply py-polars/polars/io.py:212]: If you check the `handle_projection_columns` function, it makes sure that `columns` is either None or a list of strings. That's why the `redundant-expr` error triggered!

Comment (zundertj) [reply py-polars/polars/io.py:212]: Yes, you are right, I misread the type annotation `tuple[list[int] | None, list[str] | None]`, the first is for `projection`, not `columns`.

**Review by zundertj — APPROVED** [review]

**Pull request merged**